From 26b48aab96543199fff1da5f18ffc63a4b214f42 Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 12:26:01 -0400
Subject: [ethernet] tg3: Do not include vlan acceleration features in vlan_features

Message-id: <1410179173-14304-8-git-send-email-ivecera@redhat.com>
Patchwork-id: 91663
O-Subject: [RHEL7.1 PATCH 07/19] tg3: Do not include vlan acceleration features in vlan_features
Bugzilla: 1088637
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: #1088637

Upstream commit(s):
commit 51dfe7b944998eaeb2b34d314f3a6b16a5fd621b
Author: Vlad Yasevich <vyasevic@redhat.com>
Date:   Mon Mar 24 17:52:12 2014 -0400

    tg3: Do not include vlan acceleration features in vlan_features

    Including hardware acceleration features in vlan_features breaks
    stacked vlans (Q-in-Q) by marking the bottom vlan interface as
    capable of acceleration.  This causes one of the tags to be lost
    and the packets are sent with a sing vlan header.

    Signed-off-by: Vlad Yasevich <vyasevic@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/tg3.c b/drivers/net/ethernet/broadcom/tg3.c
index 618efa8..9e535c7 100644
--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -17691,8 +17691,6 @@ static int tg3_init_one(struct pci_dev *pdev,
 
  tg3_init_bufmgr_config(tp);
 
- features |= NETIF_F_HW_VLAN_CTAG_TX | NETIF_F_HW_VLAN_CTAG_RX;
-
  /* 5700 B0 chips do not support checksumming correctly due
   * to hardware bugs.
   */
@@ -17724,7 +17722,8 @@ static int tg3_init_one(struct pci_dev *pdev,
    features |= NETIF_F_TSO_ECN;
  }
 
- dev->features |= features;
+ dev->features |= features | NETIF_F_HW_VLAN_CTAG_TX |
+    NETIF_F_HW_VLAN_CTAG_RX;
  dev->vlan_features |= features;
 
  /*
-- 
1.7.1